### ISSUE TITLE
Possiable bug in core_mmu_add_mapping()?

### ISSUE THREAD

**Issue by tiger20081015** [OP-TEE/optee_os#5201]
Hi, we met an issue related to core_mmu_add_mapping with CFG_WITH_LPAE=y in armv7 platform. The details are:

1. We place optee start address to 0xe000_0000 by setting CFG_TZDRAM_START ?= 0xe0000000 and CFG_TZDRAM_SIZE  ?= 0x02000000.  We also set CFG_RESERVED_VASPACE_SIZE  to 1G for later mapping. 
With these settings, dump_mmap_table shows below mapping table:

_E/TC:0 0 dump_mmap_table:753 type TA_RAM       va 0x3c400000..0x3dffffff pa 0xe0400000..0xe1ffffff size 0x01c00000 (pgdir)
E/TC:0 0 dump_mmap_table:753 type IO_SEC       va 0x3e000000..0x3e0fffff pa 0x00200000..0x002fffff size 0x00100000 (smallpg)
E/TC:0 0 dump_mmap_table:753 type TEE_RAM_RX   va 0xe0000000..0xe00b1fff pa 0xe0000000..0xe00b1fff size 0x000b2000 (smallpg)
E/TC:0 0 dump_mmap_table:753 type TEE_RAM_RW   va 0xe00b2000..0xe03fffff pa 0xe00b2000..0xe03fffff size 0x0034e000 (smallpg)
E/TC:0 0 dump_mmap_table:753 type SHM_VASPACE  va 0xe0400000..0xe23fffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
E/TC:0 0 dump_mmap_table:753 type IO_SEC       va 0xe2400000..0xec3fffff pa 0x01000000..0x0affffff size 0x0a000000 (pgdir)
E/TC:0 0 dump_mmap_table:753 type IO_SEC       va 0xec400000..0xfc3fffff pa 0x70000000..0x7fffffff size 0x10000000 (pgdir)
**E/TC:0 0 dump_mmap_table:753 type RES_VASPACE  va 0xfc400000..0x3c3fffff pa 0x00000000..0x3fffffff size 0x40000000 (pgdir)**_

Please note the RES_VASPACE  overwraps 4G with "va 0xfc400000..0x3c3fffff" . 

2. Then our drivers call core_mmu_add_mapping with 8M length multi times.  But we met error only after 2~3 times. Debug log shows it is caused by below code in core_mmu_add_mapping:
```
           if (core_mmu_va2idx(&tbl_info, map->va + len) >= tbl_info.num_entries) {
                                  EMSG("@@va_base:0x%lx map->va:0x%lx len:0x%x num:%d!\n", tbl_info.va_base, map->va, len, tbl_info.num_entries);
                                  return false;
           }
```
logs:
            _@@va_base:0xc0000000 map->va:0xff800000 len:0x800000 num:512!_

It seems code here doesn't take "(mmp->va + len) >= 4G" into consideration.  Maybe it should split the "map->va + len" into vritual space which belows 4G and that overwraps 4G, otherwise, CFG_RESERVED_VASPACE_SIZE=1G actually doesn't take any effect but is limited to (4G - RES_VASPACE start address).

Please give your comments and guide how to fix it if it's really a bug here. Thanks in advance!

**Comment by jenswi-linaro** [comment]
This problem doesn't seem to be specific for LPAE, you'd probably face it anyway.

I'm surprised that any mappings are placed after `TEE_RAM_RW`.

**Comment by etienne-lms** [comment]
Virtual address range spreads around the linear mapping  (`CFG_LOAD_ADDR`, here same as `CFG_TZDRAM_BASE`). `core_mmu_place_tee_ram_at_top()` tells which way it spreads (down from linear base,  or up from linear end) but the function currently only tries to fit inside a signle 1GB base block, it does not consider the size of the expect virtual space. Maybe the issue should be address here.

**Comment by tiger20081015** [comment]
@jenswi-linaro , yes, for armv7, I'd face it no matter LPAE is set or not. But for armv8 it should not exist?

@etienne-lms , yes, core_mmu_place_tee_ram_at_top() can choose which way virtual address range spreads around the linera mappping. But anyway the 4G overwrapping issue still exists.  Could the issue be handled in set_region() by splitting virtual memory address into  that belows 4G and that overwraps 4G? I made a very rough patch to fix the bug but I'm not sure if it's the best solution and maybe it would cause other issues I didn't taken into account. Could you give some advice?

diff --git a/core/arch/arm/mm/core_mmu.c b/core/arch/arm/mm/core_mmu.c
index a513e194..0cba6808 100644
--- a/core/arch/arm/mm/core_mmu.c
+++ b/core/arch/arm/mm/core_mmu.c
...
+static paddr_t _set_region(struct core_mmu_table_info *tbl_info, unsigned idx, unsigned end,
+                       uint32_t attr, paddr_t pa)
+{
+       while (idx < end) {
+               core_mmu_set_entry(tbl_info, idx, pa, attr);
+               idx++;
+               pa += 1 << tbl_info->shift;
+       }
+
+       return pa;
+}
+
+#define SZ_4G  (uint64_t)0x100000000
 static void set_region(struct core_mmu_table_info *tbl_info,
                struct tee_mmap_region *region)
 {
        unsigned end;
        unsigned idx;
        paddr_t pa;
+       size_t size;
 
        /* va, len and pa should be block aligned */
        assert(!core_mmu_get_block_offset(tbl_info, region->va));
@@ -1342,13 +1361,20 @@ static void set_region(struct core_mmu_table_info *tbl_info,
        assert(!core_mmu_get_block_offset(tbl_info, region->pa));
 
        idx = core_mmu_va2idx(tbl_info, region->va);
-       end = core_mmu_va2idx(tbl_info, region->va + region->size);
        pa = region->pa;
 
-       while (idx < end) {
-               core_mmu_set_entry(tbl_info, idx, pa, region->attr);
-               idx++;
-               pa += 1 << tbl_info->shift;
+       /* overrap 4G */
+       if ((uint64_t)region->va + region->size >= SZ_4G) {
+               /* memory space below 4G, end to 4G */
+               pa = _set_region(tbl_info, idx, tbl_info->num_entries, region->attr, pa);
+               /* memory space wrapped 4G, begin from 0x0 */
+               tbl_info->va_base = 0x0;
+               size = region->size - (size_t)(SZ_4G - (uint64_t)region->va);
+               end = size >> tbl_info->shift;
+               _set_region(tbl_info, 0, end, region->attr, pa);
+       } else {
+               end = core_mmu_va2idx(tbl_info, region->va + region->size);
+               _set_region(tbl_info, idx, end, region->attr, pa);
        }
 }

**Comment by etienne-lms** [comment]
I would not split a region VA range. Also we can map from address 0. Last, `set_region()` does not set VAs, it applies a pre-established mapping. The mapping layout of registered areas  is set from `assign_mem_va()`.

**Comment by tiger20081015** [comment]
@etienne-lms , thanks for your advice. When regarding "map from address 0", do you mean mapping secureos image or reserved va space(RES_VASPACE) from address 0?

**Comment by etienne-lms** [comment]
> @etienne-lms , thanks for your advice. When regarding "map from address 0", do you mean mapping secureos image or reserved va space(RES_VASPACE) from address 0?

Sorry, typo! I mean't **can't** of course. Low addresses are intentionally not mapped.
(i'll edit the comment)

`assign_mem_va()` uses a base VA, set from linear mapping and ASLR offset.
Currently implementation has these 2 loops to spread VAs. You're issue needs to fix there I think.

A quick way can be to consider on RES_VASPACE can be very big and consider it to select spread up/down base VA.
Another way could be to place each section above or below base VA depending on suitable space.

**Comment by jenswi-linaro** [comment]
> Another way could be to place each section above or below base VA depending on suitable space.

I believe the code is already trying to do that, hence my surprise.

**Comment by tiger20081015** [comment]
@etienne-lms , Sorry our optee version is a bit old, it still uses init_mem_map() rather than assign_mem_va() to spread the virtual memory. I just looked into the latest assign_mem_va()  function and found there's already checking code to report 4G overwrapping in line 1028, so the issue would not occur in later core_mmu_add_mapping:
```
912 static bool assign_mem_va(vaddr_t tee_ram_va,                                                                       
913 >------->------->-------  struct tee_mmap_region *memory_map) 
          ...
1027 >------->------->-------map->va = va;
1028 >------->------->------  if (ADD_OVERFLOW(va, map->size, &va))
1029 >------->------->------->-------return false;
```
According to the function,  it seems we should change CFG_TZDRAM_START  from 0xe0000000 to another place to avoid this issue.  Is there any guidance? We want to place it to near the DDR start address as much as possiable to get big RES_VASPACE. But I notice you mentioned "Low addresses are intentionally not mapped", what's the range of the unmapped low address, 0~4K?

**Comment by etienne-lms** [comment]
Right. I looked the wrong place.
The lpae implementation looks wrong for function `core_mmu_place_tee_ram_at_top()`.
I've created https://github.com/OP-TEE/optee_os/pull/5214.

**Comment by etienne-lms** [comment]
@tiger20081015, could you check #5214 fixes your issue?
